http/client.go

```diff
@@ -0,0 +1,268 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+)
+
+const (
+	maxIdleConns   = 100
+	defaultRetries = 30
+	DefaultDomain  = "localhost"
+
+	// DefaultTimeout is the default timeout for a request.
+	DefaultTimeout       = 10 * time.Second // nolint:gochecknoglobals // ok
+	defaultRetryInterval = 1 * time.Second  // nolint:gochecknoglobals // ok
+	backoffResetInterval = 10 * time.Second // nolint:gochecknoglobals // ok
+)
+
+var (
+	// ErrorUnexpectedStatusCode is used when an unexpected status code is returned.
+	// The expected status code are
+	// - 200 for a successful request
+	// - 201 for a successful creation
+	// - 202 for a successful request that is in progress
+	// - 204 for a successful request that returns no content
+	// - 400 for a request that is malformed
+	// - 404 for a request that fails due to not found
+	ErrorUnexpectedStatusCode = errors.New("unexpected status code")
+
+	defaultTransport *http.Transport = &http.Transport{
+		Proxy:               http.ProxyFromEnvironment,
+		MaxIdleConns:        maxIdleConns,
+		MaxIdleConnsPerHost: maxIdleConns,
+	} // nolint:gochecknoglobals // ok
+
+)
+
+// A Client makes HTTP requests
+//
+// Do returns an http.Response and an error.
+//
+// Except for reading the body, handlers should not modify the
+// provided Request parameters.
+type Client interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// ClientOptions are options for the client.
+// It can be used to setup a custom RoundTripper or a Timeout.
+type ClientOptions func(client *http.Client)
+
+type reqResp struct {
+	Client
+	client       *http.Client
+	baseURL      *url.URL
+	headerFields *http.Header
+	log          logr.Logger
+}
+
+// NewClient returns a new Requester given a domain name.
+// An optional http.Client can be provided.
+// An optional http.Header can be provided.
+// An optional logger can be provided.
+// An optional ClientOptions can be provided.
+func NewClient(client *http.Client, domain string, header *http.Header, log logr.Logger, opts ...ClientOptions) (Client, error) {
```

Comment:
Do you think _accept interfaces, return structs_ applies here?

Reply:
Yes!